File: code/22-Generate Parentheses/22-Generate Parentheses.py
# Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
#
# For example, given n = 3, a solution set is:
#
# [
#   "((()))",
#   "(()())",
#   "(())()",
#   "()(())",
#   "()()()"
# ]
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:

        res = []
        cur_str = ''

        def dfs(cur_str, left, right):
            """
            :param cur_str: 从根结点到叶子结点的路径字符串
            :param left: 左括号还可以使用的个数
            :param right: 右括号还可以使用的个数
            :return:
            """
            if left == 0 and right == 0:
                res.append(cur_str)
                return
            if right < left:
                return
            if left > 0:
                dfs(cur_str + '(', left - 1, right)
            if right > 0:
                dfs(cur_str + ')', left, right - 1)

        dfs(cur_str, n, n)
        return res


# class Solution:
#     def generateParenthesis(self, n):
#         if n == 0:
#             return ['']
#         ans = []
#         for c in range(n):
#             for left in self.generateParenthesis(c):
#                 for right in self.generateParenthesis(n-1-c):
#                     ans.append('({}){}'.format(left, right))
#         return ans

File: code/22-Generate Parentheses/22-Generate Parentheses.cpp
/*
LeetCode 22 - Generate Parentheses

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
*/
/*
Generate Parentheses

Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

"((()))", "(()())", "(())()", "()(())", "()()()"

*/

class Solution {
public:
    vector<string> ans;
     void rec(int left,int right,string s){

            if(left==0&&right==0){
                this->ans.push_back(s);
                return;
            }
            if(left==0){
                rec(0,right-1,s+")");
                return;
            }
            if(right==0||left>right)
                return;

            rec(left-1,right,s+"(");
            rec(left,right-1,s+")");

        }
    vector<string> generateParenthesis(int n) {

        rec(n,n,"");
        return this->ans;


    }
};
